subprojects/core/src/main/java/org/gradle/api/internal/changedetection/state/FileZipInput.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.changedetection.state;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.UncheckedIOException;
+import java.util.Enumeration;
+import java.util.function.Supplier;
+import java.util.zip.ZipFile;
+
+class FileZipInput extends ZipInput {
+
+    private final ZipFile file;
+    private final Enumeration<? extends java.util.zip.ZipEntry> entries;
+
+    public FileZipInput(File file) throws IOException {
+        this.file = new ZipFile(file);
+        this.entries = this.file.entries();
+    }
+
+    @Nullable
+    @Override
+    public ZipEntry getNextEntry() {
+        if (entries.hasMoreElements()) {
+            final java.util.zip.ZipEntry zipEntry = entries.nextElement();
+            return new JdkZipEntry(zipEntry, new Supplier<InputStream>() {
```

Comment:
Anonymous classes hold reference to outer state (which can cause leaks and occupy more memory).
If you think this could be a problem, please consider extracting this as an inner static class, like:
```java
private static class InputStreamSupplier implements Supplier<InputStream> {
    private final ZipFile file;
    private final java.util.zip.ZipEntry zipEntry;

    public InputStreamSupplier(ZipFile file, java.util.zip.ZipEntry zipEntry) {
        this.file = file;
        this.zipEntry = zipEntry;
    }

    @Override
    public InputStream get() {
        try {
            return file.getInputStream(zipEntry);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }
}
```

Reply:
Not a problem here, since the zip file is held open anyway while you're working with the entries.